### ISSUE TITLE
Feature Request: steal backing array in flat map/set

### ISSUE THREAD

**Issue by yonik** [greg7mdp/gtl#22]
One feature that would greatly decrease peak memory usage in my use-case is to have the ability to "steal" the backing array from the flat-maps (after compacting them into a "normal" array first of course.)

The use-case is currently implemented as follows: 
1) add / update many objects in a flat_hash_set
2) allocate a new array, iterate over the flat_hash_set, adding all to the array
3) delete the flat_hash_set
3) sort the array of objects

Perhaps something along the lines of:
```
// Removes and returns an array of all entries in the set.  Optimized implementations may re-use an internal array, thus
// lowering memory usage.  Only the first size() entries are defined.  It is the callers responsibility to delete the returned array.
value_type* remove_all_entries()
```

Oh, and thank you for the inspired lazy_emplace()! It's still more powerful than other map/set implementations of try_emplace since key creation can be deferred or customized as well.

**Comment by greg7mdp** [comment]
Thanks for the kind words!

> have the ability to "steal" the backing array from the flat-maps (after compacting them into a "normal" array first of course.)

Should not be too difficult to do. Let me look into it.

**Comment by greg7mdp** [comment]
Hum, there is the issue of the allocator used for allocating the array inside the flat_hash_set. If you steal the buffer, then you also need to use the same allocator to free the memory buffer eventually. I wish there was a way to construct a std::vector from a pointer, size, capacity and optional allocator.

**Comment by yonik** [comment]
> Hum, there is the issue of the allocator used for allocating the array inside the flat_hash_set

Yeah, I figured it could be a super-low-level expert method where the caller may even be in charge of calling individual destructors (making it possible is better than nothing at all IMO).  Either that, or as you suggest, your own little vector-like class.

**Comment by greg7mdp** [comment]
OK, I did the first part, added `gtl::vector`.

**Comment by h-2** [comment]
I am also interested in this! `std::flat_map` and `ankerl::unordered_dense::map` both provide `container_type extract() &&` and `void replace()` functions to access the underlying data structures.

I am not sure what the internal layout is in gtl and whether something like this would be possible.
